### ISSUE TITLE
Graph nodes having NULL as the code location

### ISSUE THREAD

**Issue by and89271** [Fraunhofer-AISEC/cpg#1009]
I am trying to find nodes from code by matching the location of the code with the location information that is contained in the nodes. 
For that i need the startLine (and startColumn). 
As i understand it every "code" node should hold the information about the file it is in (artifact), its startLine and its startColumn
My problem is that most of the nodes don't have this information.

## Example / Steps to reproduce

I have this main.cpp:

```cpp
#include <array>

struct Buffer {
    std::array<int, 10> buf;

    void set(int idx, int value) {
        buf[idx] = value;
    }
};

static void foo() {
    Buffer b_good;
    b_good.set(9, 0);
}

static void bar() {
    Buffer b_bad;
    b_bad.set(10, 0);
}

int main(){
    foo();
    bar();
}
```

I have created a compile json containing only the main.cpp file:

```json
[
    {
        "file": "/home/andreas/masterthesis/sca-training/src/challenges/buffer/main.cpp"
    }
]

```

I created the graph with neo4j with this command:

```bash
Java_OPTS="Xss1024m" cpg-neo4j --host neo4j --json-compilation-database test_buffer.json 
```

I can see the graph in my browser and issue the following query:

```sql
MATCH(n:Node) WHERE  n.startLine is null  RETURN n.name, n.artifact, n.startLine, n.startColumn, n.isImplicit
```

and get the following result:

![nullLocation](https://user-images.githubusercontent.com/120388438/207070158-c26bdb06-93b9-46a1-9f37-0346d81a5ce4.png)

**Comment by oxisto** [comment]
Update: Out of 45 AST nodes in the example, 11 do not have any location. All of those are implicit return statements which are added into the graph and were not part of the original AST (therefore, currently they do not have any code location).

**Comment by peckto** [comment]
I just re-run the example with the current main (eaec7dc19538906fdc95b51d79c907f09f4aaf21)

Filtering `type` nodes and `implicit` nodes, it looks like the following:
```
MATCH(n:Node) WHERE not n.isImplicit and n.startLine is null and not any(type IN labels(n) WHERE type =~ ".*Type") RETURN labels(n), n.code, n.name, n.artifact, n.startLine, n.startColumn, n.isImplicit, n.isInferred
```
![image](https://user-images.githubusercontent.com/2060591/210719687-17e92ce1-4c48-4b8b-a8ed-4ce5c1742568.png)

For me, it looks like the location is only missing in the `DeclaredReferenceExpression` (as `Component` is not an AST node).

**Comment by oxisto** [comment]
> I just re-run the example with the current main ([eaec7dc](https://github.com/Fraunhofer-AISEC/cpg/commit/eaec7dc19538906fdc95b51d79c907f09f4aaf21))
> 
> Filtering `type` nodes and `implicit` nodes, it looks like the following:
> 
> ```
> MATCH(n:Node) WHERE not n.isImplicit and n.startLine is null and not any(type IN labels(n) WHERE type =~ ".*Type") RETURN labels(n), n.code, n.name, n.artifact, n.startLine, n.startColumn, n.isImplicit, n.isInferred
> ```
> 
> ![image](https://user-images.githubusercontent.com/2060591/210719687-17e92ce1-4c48-4b8b-a8ed-4ce5c1742568.png)
> 
> For me, it looks like the location is only missing in the `DeclaredReferenceExpression` (as `Component` is not an AST node).

Can you see whether these are only declared references expressions within member calls? Then #1010 should fix it.

**Comment by peckto** [comment]
> Can you see whether these are only declared references expressions within member calls? Then https://github.com/Fraunhofer-AISEC/cpg/pull/1010 should fix it.

@oxisto I'm not sure if I understood you correctly.
In the given example, for all `(m:MemberCallExpression)-[:MEMER]->(p)`, `p` is always a `DeclaredReferenceExpression`.
But in general, this seems not to be true.
I found two situations, where this is not the case:

1) `(m:MemberCallExpression)-[:MEMER]->(p:ArraySubscriptionExpression)`
```c++
struct C {
        void (C::*invocation[2])();

        void foo() {
                ((*this).*(invocation[1])) ();
        }
};
```

2) `(m:MemberCallExpression)-[:MEMER]->(p:MemberCallExpression)`
(I still need to investigate)

Different question is, if the cpg for (1) and (2) is correct.

**Comment by oxisto** [comment]
> > Can you see whether these are only declared references expressions within member calls? Then #1010 should fix it.
> 
> @oxisto I'm not sure if I understood you correctly. In the given example, for all `(m:MemberCallExpression)-[:MEMER]->(p)`, `p` is always a `DeclaredReferenceExpression`. 

I meant if all `DeclaredReferenceExpression` nodes that have no location are the child node of a `MemberCallExpression` in the example

> But in general, this seems not to be true. I found two situations, where this is not the case:
> 
> 1. `(m:MemberCallExpression)-[:MEMER]->(p:ArraySubscriptionExpression)`
> 
> ```c++
> struct C {
>         void (C::*invocation[2])();
> 
>         void foo() {
>                 ((*this).*(invocation[1])) ();
>         }
> };
> ```
> 
>  2. `(m:MemberCallExpression)-[:MEMER]->(p:MemberCallExpression)`
>    (I still need to investigate)
> 
> Different question is, if the cpg for (1) and (2) is correct.

Yes this is a different question and will be handled in the new call resolver (some day)

**Comment by peckto** [comment]
> I meant if all DeclaredReferenceExpression nodes that have no location are the child node of a MemberCallExpression in the example or if they belong to some other parent node.

Ah ok, I thought, I would have written this in my comment, but apparently it was just in my head.
Yes, there are two `DeclaredReferenceExpressions` without location and both are child node of a `MemberCallExpression`.
Both are fixed with #1010.
